File: frontend/apps/web/src/services/websocketService.js
// File: src/services/websocketService.js

class WebSocketService {
  constructor() {
    this.ws = null;
    this.userId = null;
    this.token = null;
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 5;
    this.reconnectDelay = 1000;
    this.listeners = new Map();
    this.isConnecting = false;
    
    // Simple event emitter implementation
    this.events = {};
  }

  // Simple event emitter methods
  on(event, callback) {
    if (!this.events[event]) {
      this.events[event] = [];
    }
    this.events[event].push(callback);
  }

  off(event, callback) {
    if (this.events[event]) {
      this.events[event] = this.events[event].filter(cb => cb !== callback);
    }
  }

  emit(event, ...args) {
    if (this.events[event]) {
      this.events[event].forEach(callback => callback(...args));
    }
  }

  // Initialize WebSocket connection
  async connect() {
    if (this.isConnecting || (this.ws && this.ws.readyState === WebSocket.OPEN)) {
      return;
    }

    this.isConnecting = true;

    try {
      // Get token from localStorage
      this.token = localStorage.getItem('accessToken');
      if (!this.token) {
        console.error('No access token found');
        this.isConnecting = false;
        return;
      }

      // Extract user ID from JWT token
      this.userId = this.extractUserIdFromToken(this.token);
      if (!this.userId) {
        console.error('Could not extract user ID from token');
        this.isConnecting = false;
        return;
      }

      // Create WebSocket connection
      const wsUrl = `ws://localhost:8585/notifications/ws/updates/${this.userId}?token=${this.token}`;
      console.log('Connecting to WebSocket:', wsUrl);
      console.log('User ID:', this.userId);
      console.log('Token length:', this.token.length);
      
      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
        this.isConnecting = false;
        this.reconnectAttempts = 0;
        this.emit('connected');
      };

      this.ws.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);
          this.handleMessage(data);
        } catch (error) {
          console.error('Error parsing WebSocket message:', error);
        }
      };

      this.ws.onclose = (event) => {
        console.log('WebSocket disconnected:', event.code, event.reason);
        this.isConnecting = false;
        this.emit('disconnected', { code: event.code, reason: event.reason });
        
        // Attempt to reconnect if not a manual close
        if (event.code !== 1000 && this.reconnectAttempts < this.maxReconnectAttempts) {
          this.scheduleReconnect();
        }
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
        console.error('WebSocket readyState:', this.ws.readyState);
        this.isConnecting = false;
        this.emit('error', error);
      };

    } catch (error) {
      console.error('Failed to connect WebSocket:', error);
      this.isConnecting = false;
    }
  }

  // Extract user ID from JWT token
  extractUserIdFromToken(token) {
    try {
      console.log('Extracting user ID from token:', token.substring(0, 20) + '...');
      const parts = token.split('.');
      if (parts.length !== 3) {
        console.error('Invalid JWT token format');
        return null;
      }
      
      const payload = parts[1];
      const paddedPayload = payload.padEnd((payload.length + 3) & ~3, '=');
      const decodedPayload = atob(paddedPayload);
      const payloadJson = JSON.parse(decodedPayload);
      
      console.log('JWT payload:', payloadJson);
      const userId = payloadJson.sub ? parseInt(payloadJson.sub) : null;
      console.log('Extracted user ID:', userId);
      
      return userId;
    } catch (error) {
      console.error('Error extracting user ID from token:', error);
      return null;
    }
  }

  // Handle incoming WebSocket messages
  handleMessage(data) {
    console.log('WebSocket message received:', data);
    
    // Handle different message types
    if (data.type === 'new_message') {
      this.emit('new_message', data);
    } else if (data.type === 'notification') {
      this.emit('notification', data);
    } else if (data.type === 'issue_update') {
      this.emit('issue_update', data);
    } else {
      // Generic message
      this.emit('message', data);
    }
  }

  // Schedule reconnection
  scheduleReconnect() {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);
    
    console.log(`Scheduling reconnect attempt ${this.reconnectAttempts} in ${delay}ms`);
    
    setTimeout(() => {
      if (this.ws && this.ws.readyState === WebSocket.CLOSED) {
        this.connect();
      }
    }, delay);
  }

  // Send message through WebSocket
  sendMessage(message) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    } else {
      console.warn('WebSocket not connected, cannot send message');
    }
  }

  // Disconnect WebSocket
  disconnect() {
    if (this.ws) {
      this.ws.close(1000, 'Manual disconnect');
      this.ws = null;
    }
    this.isConnecting = false;
    this.reconnectAttempts = 0;
  }

  // Event listener management
  on(event, callback) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    this.listeners.get(event).push(callback);
  }

  off(event, callback) {
    if (this.listeners.has(event)) {
      const callbacks = this.listeners.get(event);
      const index = callbacks.indexOf(callback);
      if (index > -1) {
        callbacks.splice(index, 1);
      }
    }
  }

  emit(event, data) {
    if (this.listeners.has(event)) {
      this.listeners.get(event).forEach(callback => {
        try {
          callback(data);
        } catch (error) {
          console.error('Error in WebSocket event callback:', error);
        }
      });
    }
  }

  // Get connection status
  get isConnected() {
    return this.ws && this.ws.readyState === WebSocket.OPEN;
  }

  // Get user ID
  get currentUserId() {
    return this.userId;
  }
}

// Create singleton instance
const websocketService = new WebSocketService();

export default websocketService;

File: frontend/apps/mobile/mobile/lib/core/services/websocket_service.dart
// File: lib/core/services/websocket_service.dart

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/io.dart';
import '../api/api_client.dart';
import 'storage_service.dart';

class WebSocketService {
  final _storageService = StorageService();
  WebSocketChannel? _channel;
  StreamController<Map<String, dynamic>>? _messageController;
  Timer? _reconnectTimer;
  bool _isConnected = false;
  int? _userId;

  Stream<Map<String, dynamic>> get messageStream {
    _messageController ??= StreamController<Map<String, dynamic>>.broadcast();
    return _messageController!.stream;
  }

  Future<void> connect() async {
    try {
      final sessionData = await _storageService.getSession();
      if (sessionData == null) {
        print('No session data found, cannot connect to WebSocket');
        return;
      }

      // Extract user ID from session or token
      final token = await _storageService.getToken();
      if (token == null) {
        print('No token found, cannot connect to WebSocket');
        return;
      }

      // Extract user ID from JWT token
      _userId = _extractUserIdFromToken(token);
      if (_userId == null) {
        print('Could not extract user ID from token');
        return;
      }

      final wsUrl = dio.options.baseUrl.replaceFirst('http', 'ws') + 'notifications/ws/updates/$_userId?token=$token';
      
      _channel = IOWebSocketChannel.connect(wsUrl);
      _isConnected = true;

      _channel!.stream.listen(
        (data) {
          try {
            final message = json.decode(data);
            _messageController?.add(message);
          } catch (e) {
            print('Error parsing WebSocket message: $e');
          }
        },
        onError: (error) {
          print('WebSocket error: $error');
          _isConnected = false;
          _scheduleReconnect();
        },
        onDone: () {
          print('WebSocket connection closed');
          _isConnected = false;
          _scheduleReconnect();
        },
      );

      print('WebSocket connected successfully');
    } catch (e) {
      print('Failed to connect to WebSocket: $e');
      _scheduleReconnect();
    }
  }

  int? _extractUserIdFromToken(String token) {
    try {
      // JWT tokens have 3 parts separated by dots
      final parts = token.split('.');
      if (parts.length != 3) return null;
      
      // Decode the payload (second part)
      final payload = parts[1];
      // Add padding if needed
      final paddedPayload = payload.padRight((payload.length + 3) & ~3, '=');
      
      // Decode base64
      final decodedBytes = base64Url.decode(paddedPayload);
      final decodedString = utf8.decode(decodedBytes);
      
      // Parse JSON
      final payloadJson = json.decode(decodedString);
      
      // Extract user ID from 'sub' field
      final userId = payloadJson['sub'];
      return userId != null ? int.tryParse(userId.toString()) : null;
    } catch (e) {
      print('Error extracting user ID from token: $e');
      return null;
    }
  }

  void _scheduleReconnect() {
    if (_reconnectTimer?.isActive == true) return;
    
    _reconnectTimer = Timer(const Duration(seconds: 5), () {
      if (!_isConnected) {
        print('Attempting to reconnect WebSocket...');
        connect();
      }
    });
  }

  void disconnect() {
    _reconnectTimer?.cancel();
    _channel?.sink.close();
    _isConnected = false;
    print('WebSocket disconnected');
  }

  bool get isConnected => _isConnected;

  void sendMessage(String message) {
    if (_isConnected && _channel != null) {
      _channel!.sink.add(message);
    }
  }
}
